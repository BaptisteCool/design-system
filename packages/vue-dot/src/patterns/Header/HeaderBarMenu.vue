<template>
	<VMenu
		v-model="menu"
		content-class="menu-active-el"
		transition="scale-transition"
		origin="top center"
		offset-y
		left
		@input="$emit('change', $event)"
	>
		<template
			v-if="actionsList.length || loggedIn || !hideUserIcon || accountText"
			#activator="{ on }"
		>
			<VBtn
				class="menu-el text-none px-4 text-right"
				:height="buttonHeight"
				:large="!$vuetify.breakpoint.xs"
				text
				@click="$emit('click:menu')"
				v-on="on"
			>
				<VLayout
					tag="span"
					column
				>
					<!-- agent firstname and lastname or account text    -->
					<div
						v-if="!$vuetify.breakpoint.xs"
						class="subtitle-1"
					>
						{{ loggedIn ? agent : accountText }}
					</div>
					<!-- informations -->
					<div
						v-if="info || !$vuetify.breakpoint.xs"
						class="body-2"
					>
						{{ info }}
					</div>
				</VLayout>

				<!-- icon account when user is logged in -->
				<VIcon
					v-if="(!hideUserIcon && loggedIn) || ($vuetify.breakpoint.xs && !loggedIn)"
					size="32px"
					color="grey darken-1"
					class="ml-2 pa-1 round-icon"
				>
					{{ userIcon }}
				</VIcon>
			</VBtn>
		</template>
		<VList
			:width="$vuetify.breakpoint.xs ? 360 : null"
			class="py-0 subtitle-1"
		>
			<!-- list of optional actions -->
			<VListItem
				v-for="(item, index) in actionsList"
				:key="index"
				@click="$emit('click:action', index)"
			>
				<VListItemContent>
					{{ item }}
				</VListItemContent>
			</VListItem>
			<!-- logout -->
			<VListItem
				v-if="loggedIn"
				@click="$emit('click:logout')"
			>
				<VIcon
					color="primary"
					class="mr-2"
				>
					{{ logoutIcon }}
				</VIcon>

				{{ locales.logOut }}
			</VListItem>
		</VList>
	</VMenu>
</template>

<script lang="ts">
	import Vue from 'vue';
	import locales from './locales';
	import Component from 'vue-class-component';

	import { mapActions, mapState } from 'vuex';

	import { mdiAccount, mdiExitToApp } from '@mdi/js';

	const Props = Vue.extend({
		props: {
			loggedIn: {
				type: Boolean,
				default: false
			},
			firstname: {
				type: String,
				default: null
			},
			hideUserIcon: {
				type: Boolean,
				default: false
			},
			accountText: {
				type: String,
				default: locales.account
			},
			lastname: {
				type: String,
				default: null
			},
			info: {
				type: String,
				default: null
			},
			actionsList: {
				type: [Array, Object],
				default: () => []
			}
		}

	});

	/** The profile button in the Header */
	@Component<HeaderMenu>({

	})
	export default class HeaderMenu extends Props {
		// Icons
		userIcon = mdiAccount;
		logoutIcon = mdiExitToApp;
		locales = locales;
		width: number = 0;
		menu = false;

		get agent() {
			if (this.firstname && this.lastname) {
				return `${this.firstname} ${this.lastname}`;
			}
			return null;
		}

		get buttonHeight() {
			if (this.$vuetify.breakpoint.xs) {
				return 32;
			} else {
				return this.info ? 71 : 55;
			}
		}
	}
</script>

<style lang="scss" scoped>
	.menu-active-el {
		border-radius: 4px;
	}

	.menu-el {
		right: 15px;
		position: absolute;
		text-align: left;
	}

	.round-icon {
		border-radius: 50%;
	}

	@media only screen and (max-width: 600px) {
		.menu-el {
			position: relative;
			margin-top: 10px;
			right: 0;
		}
	}
</style>
