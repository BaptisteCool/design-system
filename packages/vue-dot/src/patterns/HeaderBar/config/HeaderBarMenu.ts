export default {
	menu: {
		contentClass: 'menu-active-el',
		transition: 'scale-transition',
		origin: 'top center',
		offsetY: true,
		left: true
	},
	button: {
		class: 'menu-el text-none px-4 text-right no-text-transform',
		text: true
	},
	buttonText: {
		class: 'subtitle-1 text-truncate'
	},
	buttonInfo: {
		class: 'body-2 text-truncate'
	},
	buttonUserIcon: {
		color: 'grey darken-1',
		class: 'round-icon ml-1'
	}
};
