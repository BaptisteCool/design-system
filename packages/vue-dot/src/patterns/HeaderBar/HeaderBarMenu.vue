<template>
	<VMenu
		v-if="buttonText || showIcon"
		v-model="menu"
		v-bind="options.menu"
		@input="$emit('change', $event)"
	>
		<template #activator="{ on }">
			<VBtn
				data-test="buttonMenu"
				v-bind="options.button"
				:height="buttonHeight"
				:large="!$vuetify.breakpoint.smAndDown"
				@click="clickMenu"
				v-on="on"
			>
				<VLayout
					v-if="buttonText"
					class="btn-text"
					column
				>
					<!-- agent firstname and lastname or account text  -->
					<div v-bind="options.buttonText">
						{{ buttonText }}
					</div>
					<!-- informations -->
					<div
						v-if="info"
						v-bind="options.buttonInfo"
					>
						{{ info.trim() }}
					</div>
				</VLayout>

				<!-- icon account
				when user is logged in and icon not hidden
				when mobile version and the user is not logged In
				-->
				<VIcon
					v-if="showIcon"
					v-bind="options.buttonUserIcon"
				>
					{{ mdiAccount }}
				</VIcon>
			</VBtn>
		</template>
		<HeaderBarActions
			:actions-list="actionsList"
			:logged-in="loggedIn"
			@click:action="clickAction"
			@click:logout="clickLogout"
		/>
	</VMenu>
</template>

<script lang="ts">
	import HeaderBarActions from './HeaderBarActions.vue';
	import Vue from 'vue';
	import Component from 'vue-class-component';

	import customizable, { Options } from '../../mixins/customizable';
	import config from './config/HeaderBarMenu';

	import { mapActions, mapState } from 'vuex';

	import { mdiAccount } from '@mdi/js';

	import header from './mixins/header';

	/** The profile button in the Header */
	@Component<HeaderBarMenu>({
		components: { HeaderBarActions },
		mixins: [
			// Default configuration
			customizable(config),
			header
		]
	})
	export default class HeaderBarMenu extends Vue {
		// Mixin computed data
		options!: Options;

		// mixins
		actionsList!: [string];
		firstname!: string;
		lastname!: string;
		loggedIn!: boolean;
		accountText!: string;
		hideUserIcon!: boolean;
		info!: string;

		// Icons
		mdiAccount = mdiAccount;

		menu = false;

		// compose the user display name
		get agent() {
			const firstname =
				typeof this.firstname === 'string' ? this.firstname.trim() : '';
			const lastname =
				typeof this.lastname === 'string' ? this.lastname.trim() : '';
			if (firstname.length > 0 || lastname.length > 0) {
				return `${firstname} ${lastname}`;
			}
			return null;
		}

		/** calcul the button height */
		get buttonHeight() {
			if (this.$vuetify.breakpoint.smAndDown) {
				return 32;
			} else {
				return this.info ? 71 : 55;
			}
		}

		// calcul the button text to display
		get buttonText() {
			if (this.$vuetify.breakpoint.smAndDown) {
				return null;
			}

			if (this.loggedIn) {
				return this.agent;
			} else if (this.accountText && this.accountText.length > 0) {
				return this.accountText.trim();
			}

			return null;
		}

		/** calcul if show the icon */
		get showIcon() {
			return (
				(!this.hideUserIcon &&
					this.loggedIn &&
					!this.$vuetify.breakpoint.smAndDown) ||
				(this.$vuetify.breakpoint.smAndDown && !this.loggedIn)
			);
		}
	}
</script>

<style lang="scss" scoped>
	.btn-text {
		max-width: 270px;
	}

	.menu-active-el {
		border-radius: 4px;
	}

	.menu-el {
		right: 15px;
		position: absolute;
		text-align: left;
	}

	.round-icon {
		border-radius: 50%;
	}

	@media only screen and (max-width: 600px) {
		.menu-el {
			position: relative;
			margin-top: 10px;
			right: 0;
		}
	}
</style>
