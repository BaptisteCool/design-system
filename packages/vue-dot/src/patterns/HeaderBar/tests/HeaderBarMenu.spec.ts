import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';
import html from '@/tests/html';

import HeaderBarMenu from '../HeaderBarMenu.vue';

let wrapper: Wrapper<Vue>;

// Tests
describe('HeaderBarMenu test', () => {
	it('renders connected correctly', () => {
		// Mount component
		wrapper = mountComponent(HeaderBarMenu, {
			propsData: {
				firstname: 'Prénom',
				lastname: 'nom',
				loggedIn: true,
				actionsList: ['text1', 'text2'],
				info: 'information supplémentaire'
			}
		}, true);

		expect(html(wrapper)).toMatchSnapshot();

		// click menu
		wrapper.find('[data-test="buttonMenu"]').trigger('click');

		// click action 0 text1
		wrapper.find('[data-test="itemAction-0"]').trigger('click');

		// logout
		wrapper.find('[data-test="itemLogout"]').trigger('click');
	});

	it('renders not connected', () => {
		// Mount component
		wrapper = mountComponent(HeaderBarMenu, {
			propsData: {
				actionsList: ['text1', 'text2'],
				firstname: 'Prénom',
				lastname: 'nom',
				info: 'information supplémentaire'
			}
		});

		expect(wrapper.text()).not.toBe('Prénom');
		expect(wrapper.text()).not.toBe('nom');
		expect(html(wrapper)).toMatchSnapshot();
	});
});
