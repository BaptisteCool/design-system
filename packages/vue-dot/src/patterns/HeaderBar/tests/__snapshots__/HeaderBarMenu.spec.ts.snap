// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBarMenu test renders connected correctly 1`] = `
<div class="v-menu"><button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--large menu-el text-none px-4 text-right no-text-transform" style="height: 71px;" data-test="buttonMenu"><span class="v-btn__content"><div class="layout btn-text column"><div class="subtitle-1 text-truncate">
					Prénom nom
				</div> <div class="body-2 text-truncate">
					information supplémentaire
				</div></div> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="24" width="24" role="img" aria-hidden="true" class="v-icon notranslate v-icon--svg theme--light grey--text text--darken-1 round-icon ml-1"><path d="M12,4A4,4 0 0,1 16,8A4,4 0 0,1 12,12A4,4 0 0,1 8,8A4,4 0 0,1 12,4M12,14C16.42,14 20,15.79 20,18V20H4V18C4,15.79 7.58,14 12,14Z"></path></svg></span></button>
  <div role="menu" class="v-menu__content theme--light menu-active-el" style="max-height: auto; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top center; z-index: 0; display: none;"></div>
</div>
`;

exports[`HeaderBarMenu test renders not connected 1`] = `
<vmenu-stub opendelay="0" closedelay="0" contentclass="menu-active-el" left="true" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top center" transition="scale-transition">
  <headerbaractions-stub accounttext="Mon compte" actionslist="text1,text2" navigationlist=""></headerbaractions-stub>
</vmenu-stub>
`;
